Added user retrieval

Slartibartfass2 opened the pull request:
Closes #32 

## What I have made

This waits for #44 

I added the implementation for the following calls
- `getUserById`
- `getUserByEmail`
- `getAllUsers`

The code is straightforward and designed to be reviewed commit-by-commit.

## Checklist

Either tick or cross out the items that do not apply (using \~\~example text\~\~) and give a reason why the item does not apply.

- [x] I have updated the documentation accordingly and commented my code
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] (for reviewer) I have checked the implementation against the requirements

Slartibartfass2 committed:
feat: add repository implementation for getUserById

Slartibartfass2 committed:
feat: add service implementation for getUserById

Slartibartfass2 committed:
test: add test helper for invalid result assertions

Slartibartfass2 committed:
feat: add validation for Base.Id

Slartibartfass2 committed:
feat: add repository implementation for getUserByEmail

Slartibartfass2 committed:
feat: add service implementation for getUserByEmail

Slartibartfass2 committed:
feat: add repository implementation for getAllUsers

Slartibartfass2 committed:
feat: add service implementation for getAllUsers

Slartibartfass2 committed:
feat: add validation for Base.Nothing

Merseleo commented on src/main/kotlin/model/SnowballRException.kt:
```diff
@@ -58,4 +61,41 @@ sealed class SnowballRException(
+     *
+     * This exception is intended to provide a clear and structured way to handle
+     * scenarios where a particular entity is accessed, but the current user is not
+     * authorized to do so.
```
```suggestion
     * scenarios where a particular entity is accessed without permission.
```
Otherwise you only repeat yourself unnecessarily?

Merseleo commented on src/main/kotlin/model/SnowballRException.kt:
```diff
@@ -58,4 +61,41 @@ sealed class SnowballRException(
+     * various entities.
+     *
+     * @constructor Creates an [UnauthorizedException] with the current user's ID, the
+     * name and ID of the accessed entity.
```
```suggestion
     * type and ID of the accessed entity.
```
Is it really the name of the entity? Could be confusing, because the name of a project is actually something different, right?

Slartibartfass2 replied:
Yes, I agree. I changed for all exceptions.

Merseleo commented on src/main/kotlin/model/SnowballRException.kt:
```diff
@@ -58,4 +61,41 @@ sealed class SnowballRException(
+        ) : UnauthorizedException(currentUserId, "all $accessedEntityName.") {
+            class User(
+                currentUserId: String,
+            ) : UnauthorizedException(currentUserId, "user")
```
```suggestion
            ) : All(currentUserId, "user")
```
Why don't you extend your `All` class? Otherwise the message should be wrong or did I misunderstood these sealed classes and inheritance?

Slartibartfass2 replied:
Yes, that's correct.  This was an issue from my side.

Merseleo replied:
Ok 👍

Merseleo commented on src/main/kotlin/model/SnowballRException.kt:
```diff
@@ -58,4 +61,41 @@ sealed class SnowballRException(
+            accessedEntityName: String,
+            accessedEntityId: String,
+        ) : UnauthorizedException(currentUserId, "$accessedEntityName with ID '$accessedEntityId'.") {
+            class Project
```
And why did you only declare this class? Couldn't you implement this too?

Slartibartfass2 replied:
This must've been an artifact from development. I added the `Single` exception, but don't use it yet. We will use it for sure in other requests.

Merseleo replied:
Ok and I agree, I think we will need this exception soon enough.

Merseleo commented on src/test/kotlin/TestUtils.kt:
Nice idea to add this assertion helper and reduce duplicate code. However, I think you missed on case: `fun When an unknown request is validated, then the 'UnknownRequest' issue is returned() {` in the `ValidatorTest` or was this intentional?

Slartibartfass2 replied:
Ah yes, thank your for that.

Merseleo commented on src/test/kotlin/DataBuilder.kt:
```diff
@@ -76,4 +79,26 @@ object DataBuilder {
+        email: String = "test.email@example.com",
+        firstName: String = "Test",
+        lastName: String = "User",
+        role: UserRole = UserRole.USER_ROLE_UNSPECIFIED,
+        status: UserStatus = UserStatus.USER_STATUS_UNSPECIFIED,
```
Why no meaningful role and status?

Slartibartfass2 replied:
So that it is specifically set when it's required for a test.

Merseleo replied:
Ok, understandable.

Merseleo commented on src/main/kotlin/validation/Validator.kt:
```diff
@@ -29,6 +30,9 @@ fun <T> validateRequest(request: T): EitherNel<ValidationIssue, Unit> =
         is CriterionOuterClass.Criterion.Create -> CriterionValidator.validateCreateRequest(request)
+        // Base
+        is Base.Id -> BaseValidator.validateId(request)
+        is Base.Nothing -> Either.Right(Unit)
```
Only question of interest? Why did you added these checks to the end of this switch?!? statement?

Slartibartfass2 replied:
Yes, this is a Kotlin switch statement.

Is this question about the position of the checks in the whole statement? If so, then there isn't a higher reason behind that.

Merseleo replied:
Yes that was a question about the position and ok, it's fine if they are at the end. The order of the statements should not matter.

Merseleo commented on src/main/kotlin/validation/BaseValidator.kt:
```diff
@@ -0,0 +1,16 @@
+import snowballr.Base
+
+/**
+ * An object that validates the [Base] requests.
```
```suggestion
 * An validator for the [Base] requests.
```

Slartibartfass2 replied:
I adjusted the comment for all validators.

Merseleo replied:
Thank you, missed the remaining comments.

Merseleo commented on src/main/kotlin/validation/BaseValidator.kt:
```diff
@@ -0,0 +1,16 @@
+object BaseValidator {
+    fun validateId(request: Base.Id): EitherNel<ValidationIssue, Unit> =
+        either {
+            ensureFieldNonBlank("id", request.id)
```
I didn't keep up with the last discussion: Do we now only use UUID as global IDs? If so, why is it not checked in the validation whether it is a valid UUID? If not, why don't we use it and is there perhaps a more precise specification than the ID must not be empty?

Slartibartfass2 replied:
Currently, we have a mix. Most entities use a UUID, others such as the paper and the project use a numerical ID. As this request class is very generic and frequently used, I wouldn't make any assumptions about the format of the ID and validate this at a later point.

Merseleo replied:
Ok, but does it make sense to use numerical Ids for the paper and projects? Why not using UUIDs as well and then check for the format in this validator? Because if the ID of the Paper is 1232445 or a UUID doesn't matter imo and for projects I kind of understand, but even in this case: is it really necessary to know to number of already created projects? Or have numerical Ids for projects?

Slartibartfass2 replied:
I think our initial thought was that it is easier to display in the frontend, but I checked again and I think it is fine to use UUIDs. We'll have to figure out a way for the reading list and the global paper view, but otherwise this shouldn't be a problem.
Maybe we can use Base64 or something else to show a unique but short ID for the papers. For the project papers we already have the local ID, which is fine.

Merseleo replied:
Exactly and a UUID is not that long, so imo even displaying them directly in the global paper view should be ok.

Merseleo commented on src/main/kotlin/repository/UserTableRepo.kt:
```diff
@@ -0,0 +1,64 @@
+ * abstraction over the underlying database implementation. By using this interface, the logic
+ * for creating and managing users can remain decoupled from the specifics of the database layer.
+ */
+interface IUserTableRepo {
```
Maybe add a notice, that the behavior of these operations is documented in the API documentation, where the RPCs have the same name?

Merseleo replied:
It's the same in the `UserService.kt`. Or did you mention this in the Wiki?

Slartibartfass2 replied:
The repositories should be independent of the API and should only provide CRUD operations for the entity they represent.

Slartibartfass2 replied:
I added comments for all service methods.

Merseleo commented on src/main/kotlin/service/UserService.kt:
```diff
@@ -0,0 +1,67 @@
+        repo.getUserById(id.id).toGrpcUser()
+
+    override suspend fun getUserByEmail(email: Base.Id): UserOuterClass.User {
+        // TODO: move this validation to validation layer by using separate Email type
```
Why can't you just implement this? Do we need to release the new API version first? If there is no specific reason, please move it to the validation layer.

Merseleo replied:
And the `IllegalArgumentExpection` should then become a `InvalidArgumentException`, right?

Slartibartfass2 replied:
Yes, I am currently waiting for the API release. This should be done before we're finished here.

There's no such thing as an `InvalidArgumentException`, but this will resolve either way.

Merseleo replied:
Ok, yes I hope that Luca will finish this soon so we can release tomorrow. 

Exactly, for now there was no `InvalidArgumentException` but for an invalid email format we need this exception.

Merseleo replied:
Because when I request a user with an email, that is not valid, I get the following error: "INVALID_ARGUMENT" and nothing more.

Slartibartfass2 replied:
As soon as the new API version is released, this will move to the input validation layer. There we don't throw an InvalidArgumentException but return the status code directly.

Slartibartfass2 replied:
That no error message is thrown is because we catch the exception and only return the error code. As this exception might be thrown by another part of the program (and not deliberately from us), sending the exception message would share the internal structure with the usage, which is bad.

Merseleo replied:
Ok.

Merseleo commented on src/main/kotlin/service/UserService.kt:
```diff
@@ -0,0 +1,67 @@
+    private val repo: IUserTableRepo,
+) : IUserService {
+    override suspend fun getUserById(id: Base.Id): UserOuterClass.User =
+        // TODO: think about who can request each user by ID
```
I thought we agreed that only other users from the same project and the super admin can request another user? Because this is necessary as we need to fetch information about who reviewed what paper. Can you implement this check? Or do you want to discuss it first and deal with this later? In this case, please create an issue as it is difficult to find TODO comments later.

Slartibartfass2 replied:
Alright, I wasn't aware of that. This requires a lot of fetching from different tables, so I will do this in the next review round. The weather is great today. 🌞

Merseleo replied:
We can also discuss this on Tuesday again and implement this behavior in another fix issue?

Slartibartfass2 replied:
Yes, let's do that. Maybe we can already return the required information when fetching the review data.

I will investigate how we use this in the frontend and will prepare the discussion.

Slartibartfass2 replied:
A quick update on this: We only ever use the name of the user in the frontend for this call.
I would make another call, e.g., `getUserNameById` and then restrict it to only project members or the user itself.
The full user object can only be requested by a server admin or the user itself.

What are your thoughts on that?

Merseleo replied:
In my opinion it is ok to request the entire user object as all project members probably know the email anyway and no other personal information would be leaked. But lets discuss this later.

Merseleo commented on src/main/kotlin/service/UserService.kt:
```diff
@@ -0,0 +1,67 @@
+    }
+
+    override suspend fun getAllUsers(): UserOuterClass.User.List {
+        val currentUser = repo.getUserById(dummyUserId!!)
+        // Check whether the current user has access to retrieve all users
+        // TODO: remove dummy user when user management is implemented
+        if (currentUser.role != UserRole.USER_ROLE_ADMIN) {
+            throw UnauthorizedException.All.User(dummyUserId!!)
+        }
+
+        val users = repo.getAllUsers()
+
+        val builder = UserOuterClass.User.List.newBuilder()
+        users.forEach { builder.addUsers(it.toGrpcUser()) }
+        return builder.build()
```
And this will break our frontend as we can't request suggestions for invitations anymore, right?

Slartibartfass2 replied:
Yes, but there is [this issue](https://github.com/SE-UUlm/snowballr-api/issues/50) which will provide an alternative for the frontend.

Merseleo replied:
Ok

Merseleo commented on src/test/kotlin/service/user/GetAllUsersTest.kt:
```diff
@@ -0,0 +1,60 @@
+            coEvery { userRepoMock.getUserById(dummyUserId!!) } returns nonAdminUser
+            coEvery { userRepoMock.getAllUsers() } returns emptyList()
+
+            assertThrows<UnauthorizedException.All.User> { mainService.getAllUsers() }
```
Why did you check for the specific exception in this case, but in the other service tests you only assert that some exception is thrown?!

Slartibartfass2 replied:
Because in the other tests we mock that a call me make in this service method throws, but in this case, we produce a state in which the service method itself throws.

Merseleo replied:
Ok.

Merseleo reviewed (changes requested):
So this has taken a little longer now, because I still have to familiarise myself with the whole structure, Kotlin and everything else. It's possible that some questions or suggestions are very stupid, but that's because I'm still a bit inexperienced. In such cases, however, I am very grateful for clarification.

Apart from that: it looks very good, the tests are running and I haven't found any logical errors or anything similar. But it is also difficult to test the functionality when it's not connected to anything yet.

Slartibartfass2 committed:
docs: clarify SnowballRException's comments

Slartibartfass2 committed:
docs: add comments to service methods to which API call they refer to

Slartibartfass2 committed:
docs: improve comments for validators.

Slartibartfass2 committed:
refactor: refactor ID parsing

Slartibartfass2 reviewed (commented):
Thank you for the review. Keep in mind that there are no stupid questions.

For testing, you can use a gRPC client. I use [ezy](https://github.com/getezy/ezy).

Slartibartfass2 committed:
test: refactor test where helper can be used

Merseleo commented on src/main/kotlin/repository/UserTableRepo.kt:
```diff
@@ -37,9 +37,11 @@ class UserTableRepo(
 ) : IUserTableRepo {
     override suspend fun getUserById(id: String): User =
         db.dbQuery {
+            val uuid = parseUUID(id, "user")
```
Then again another question: When you parse the Id here and it fails and an exception is thrown: Is this exception automatically catched? Because you did not consider this case here, right?

Slartibartfass2 replied:
Every exception thrown in the service or repo layer is caught by the exception interceptor. I added a test case for this.

Merseleo replied:
Ah ok, thank you.

Merseleo commented on src/main/kotlin/table/TableHelper.kt:
```diff
@@ -14,10 +17,11 @@ import java.util.UUID
  * @param id The ID of the entity as [String].
- * @return The ID of the entity as [EntityID], or null if no entity exists or the [id] couldn't be parsed to an [Int].
+ * @return The ID of the entity as [EntityID], or null if no entity exists.
+ * @throws InvalidIdException.IntId If [id] cannot be parsed to an integer.
```
or is negative, right?

Slartibartfass2 replied:
I changed it to "integer ID" so it implies all that and doesn't require us to share all the implementation details.

Merseleo replied:
Not necessary anymore.

Merseleo commented on src/main/kotlin/repository/UserTableRepo.kt:
```diff
@@ -0,0 +1,66 @@
+                .where { UserTable.id eq uuid }
+                .map { it.toUser() }
+                .singleOrNull()
+                ?: throw NotFoundException.User(id)
```
```suggestion
                ?: throw NotFoundException.User(id, "ID")
```
Right? Because otherwise you get the following error message: "NOT_FOUND: User with null '4d7d037c-6ceb-406c-aa7e-da39823b9801' not found."

Slartibartfass2 replied:
You are completely right. Good catch! I got something wrong with the default values. It is fixed now and this is not necessary.

Merseleo replied:
Ok, I think it works now correctly.

Merseleo reviewed (changes requested):
> For testing, you can use a gRPC client. I use [ezy](https://github.com/getezy/ezy).

Good idea, thank you. I set up the HTTP Client in IntelliJ to test the calls and this helped a lot to check the implementation. 

I think you have made great progress, and I am looking forward to using these calls and the other implementations.

Slartibartfass2 committed:
feat: add check whether dummy user already exists

Otherwise, one would have to delete the DB data every time the backend is restarted.

Slartibartfass2 committed:
chore: integrate review feedback

Slartibartfass2 committed:
feat: make project and paper table UUIDTables

Slartibartfass2 reviewed (commented):
Thank you for the review :)

Slartibartfass2 committed:
chore: fix exception error message

Slartibartfass2 committed:
chore: bump API version to v0.5.0

Slartibartfass2 committed:
feat: adjust backend to v0.5.0

This includes adding the new call and adjusting the getUserByEmail call to validate the email directly in the input validation layer.

Slartibartfass2 committed:
feat: add validation for UUID fields

Merseleo commented on src/main/kotlin/model/ValidationIssue.kt:
```diff
@@ -57,3 +57,27 @@ data class EnumUnspecified(
+    val name: String,
+    val id: String,
+) : ValidationIssue {
+    override fun toString(): String = "The ID '$id' is invalid for the field '$name'."
```
Isn't this a bit unnecessary complicated: For example I got the error message "The ID '4d7d037c-6ceb-406c-aa7eda39823b9801' is invalid for the field 'id'." and I already said that my id is invalid? Do we really need to know in which field it was wrong?

Slartibartfass2 replied:
This may not be only the id field but sometimes the `project_id` and the `user_id` field. In requests, which contain more than one ID it would be better to display which of the IDs is invalid.

Merseleo replied:
Ah right, so lets keep it.

Merseleo commented on src/main/kotlin/table/TableHelper.kt:
```diff
@@ -14,28 +15,11 @@ import java.util.UUID
  */
 fun UUIDTable.getEntityId(id: String): EntityID<UUID>? {
-    val uuid = runCatching { UUID.fromString(id) }.getOrNull() ?: return null
+    val uuid = parseUUID(id, "user")
```
Why do you generally speak of 'entity' and in this statement you set the entity type to "user"?

Slartibartfass2 replied:
Good catch, that's wrong, the type is passed as a parameter and I created helper methods for the different entities as they will be used more often.

Merseleo replied:
Things are looking better now.

Merseleo commented on src/main/kotlin/validation/ValidationHelper.kt:
```diff
@@ -46,3 +48,14 @@ fun Raise<ValidationIssue>.ensureEnumNotUnspecified(
 ) = ensure(value.ordinal > 0) { EnumUnspecified(name) }
+
+/**
+ * Ensures that the provided [id] of field [name] has an invalid format.
```
```suggestion
 * Ensures that the provided [id] of field [name] has a valid format.
```
Right?

Slartibartfass2 replied:
Sometimes I wonder if I'm stupid ...
Thanks

Merseleo replied:
You are welcome

Merseleo reviewed (changes requested):
Great that you already integrated the new API version. I think there are some minor things left and the entire thing with the access control for the get requests will be done in #54.

Slartibartfass2 committed:
refactor: add getEntityId methods for single entities.

Slartibartfass2 committed:
docs: fix comment

Merseleo reviewed (approved):
This PR is done, lets start the merge train ;)
